File: composeApp/src/androidMain/kotlin/co/jarias/flexapp/ui/screens/bingo/card_scanner/BingoCardParser.kt
package co.jarias.flexapp.ui.screens.bingo.card_scanner

import com.google.mlkit.vision.text.Text

sealed class ParseResult {
    data class Success(val grid: List<List<String>>) : ParseResult()
    data class InvalidRange(val number: Int, val column: String) : ParseResult()
    object Incomplete : ParseResult()
}

class BingoCardParser {
    private val columnRanges = listOf(1..15, 16..30, 31..45, 46..60, 61..75)
    private val columnLabels = listOf("B", "I", "N", "G", "O")

    fun parse(text: Text): ParseResult {
        val candidates = mutableListOf<DetectedNumber>()

        for (block in text.textBlocks) {
            for (line in block.lines) {
                for (element in line.elements) {
                    val value = element.text.filter { it.isDigit() }
                    val num = value.toIntOrNull()
                    if (num != null) {
                        candidates.add(
                            DetectedNumber(
                                value = num,
                                x = element.boundingBox?.centerX() ?: 0,
                                y = element.boundingBox?.centerY() ?: 0,
                                height = element.boundingBox?.height() ?: 0
                            )
                        )
                    }
                }
            }
        }

        if (candidates.size < 20) return ParseResult.Incomplete

        // 1. Filter out noise based on size (Bingo numbers are usually similar in size)
        val medianHeight = candidates.map { it.height }.sorted().let {
            if (it.isEmpty()) 0 else it[it.size / 2]
        }
        val validCandidates = candidates.filter { it.height >= medianHeight * 0.6 }

        if (validCandidates.size < 20) return ParseResult.Incomplete

        // 2. Group into 5 columns
        val sortedByX = validCandidates.sortedBy { it.x }
        val columns = mutableListOf<MutableList<DetectedNumber>>()
        
        if (sortedByX.isNotEmpty()) {
            var currentColumn = mutableListOf<DetectedNumber>()
            currentColumn.add(sortedByX[0])
            columns.add(currentColumn)
            
            // We expect 5 columns. Find the gaps.
            // If we have more than 5, it means some columns are split or there's horizontal noise.
            // If less than 5, we missed a column.
            
            // A more robust way to find 5 columns:
            // Identify the 4 largest gaps in X coordinates
            val gaps = mutableListOf<Pair<Int, Double>>() // Index to Gap Size
            for (i in 1 until sortedByX.size) {
                gaps.add(i to (sortedByX[i].x - sortedByX[i-1].x).toDouble())
            }
            
            // Ensure we have enough gaps to even try
            if (gaps.size < 4) return ParseResult.Incomplete

            val topGaps = gaps.sortedByDescending { it.second }.take(4).map { it.first }.sorted()
            
            var startIdx = 0
            val partitionedColumns = mutableListOf<List<DetectedNumber>>()
            for (gapIdx in topGaps) {
                partitionedColumns.add(sortedByX.subList(startIdx, gapIdx))
                startIdx = gapIdx
            }
            partitionedColumns.add(sortedByX.subList(startIdx, sortedByX.size))
            
            if (partitionedColumns.size == 5) {
                // 3. Fill the grid
                val finalGrid = List(5) { MutableList(5) { "" } }
                for (c in 0..4) {
                    val colNumbers = partitionedColumns[c].sortedBy { it.y }
                    val expectedCount = if (c == 2) 4 else 5
                    
                    // If we have headers (B I N G O) detected as numbers, they'll be at the top.
                    // If we have noise at the bottom, it'll be at the end.
                    // We try to find the contiguous block of expectedCount that fits the range.
                    
                    var bestMatch: List<DetectedNumber>? = null
                    for (i in 0..(colNumbers.size - expectedCount)) {
                        val sub = colNumbers.subList(i, i + expectedCount)
                        val invalidCount = sub.count { it.value !in columnRanges[c] }
                        if (bestMatch == null || invalidCount < bestMatch!!.count { it.value !in columnRanges[c] }) {
                            bestMatch = sub
                        }
                    }
                    
                    if (bestMatch == null) return ParseResult.Incomplete

                    var targetIdx = 0
                    for (r in 0..4) {
                        if (r == 2 && c == 2) {
                            finalGrid[r][c] = "" // FREE
                        } else {
                            val detected = bestMatch[targetIdx++]
                            // Even if it's an error, we report it so the user sees it in the grid.
                            finalGrid[r][c] = detected.value.toString()
                        }
                    }
                }
                
                // Final validation of ranges for the whole grid
                for (c in 0..4) {
                    for (r in 0..4) {
                        if (r == 2 && c == 2) continue
                        val valStr = finalGrid[r][c]
                        val num = valStr.toIntOrNull() ?: 0
                        if (num !in columnRanges[c]) {
                            return ParseResult.InvalidRange(num, columnLabels[c])
                        }
                    }
                }

                return ParseResult.Success(finalGrid)
            }
        }

        return ParseResult.Incomplete
    }
    
    data class DetectedNumber(val value: Int, val x: Int, val y: Int, val height: Int)
}

File: iosApp/iosApp/Screens/bingo/card_scanner/BingoCardParser.swift
import Foundation
import Vision

struct DetectedNumber {
    let value: Int
    let x: CGFloat
    let y: CGFloat
    let height: CGFloat
}

enum ParseResult: Equatable {
    case success(grid: [[String]])
    case invalidRange(number: Int, column: String)
    case incomplete
}

class BingoCardParser {
    private let columnRanges: [ClosedRange<Int>] = [1...15, 16...30, 31...45, 46...60, 61...75]
    private let columnLabels = ["B", "I", "N", "G", "O"]

    /// Parse recognized text observations from Vision into a 5x5 Bingo card grid.
    /// - Parameters:
    ///   - observations: Results from VNRecognizeTextRequest
    ///   - imageSize: The size of the image being processed (in pixels)
    /// - Returns: ParseResult with the grid or error
    func parse(from observations: [VNRecognizedTextObservation], imageSize: CGSize) -> ParseResult {
        let candidates = extractNumbers(from: observations, imageSize: imageSize)

        if candidates.count < 20 { return .incomplete }

        // 1. Filter out noise based on size (Bingo numbers are usually similar in size)
        let medianHeight = candidates.map { $0.height }.sorted().median()
        let validCandidates = candidates.filter { $0.height >= medianHeight * 0.6 }

        if validCandidates.count < 20 { return .incomplete }

        // 2. Group into 5 columns based on X-coordinate gaps
        let sortedByX = validCandidates.sorted { $0.x < $1.x }
        var gaps: [(index: Int, size: CGFloat)] = []
        for i in 1..<sortedByX.count {
            gaps.append((i, sortedByX[i].x - sortedByX[i - 1].x))
        }

        if gaps.count < 4 { return .incomplete }

        let topGaps = gaps.sorted { $0.size > $1.size }
            .prefix(4)
            .map { $0.index }
            .sorted()

        var partitionedColumns: [[DetectedNumber]] = []
        var startIdx = 0
        for gapIdx in topGaps {
            partitionedColumns.append(Array(sortedByX[startIdx..<gapIdx]))
            startIdx = gapIdx
        }
        partitionedColumns.append(Array(sortedByX[startIdx..<sortedByX.count]))

        if partitionedColumns.count != 5 { return .incomplete }

        // 3. Fill the grid
        var finalGrid: [[String]] = Array(repeating: Array(repeating: "", count: 5), count: 5)

        for c in 0..<5 {
            let colNumbers = partitionedColumns[c].sorted { $0.y < $1.y }
            let expectedCount = (c == 2) ? 4 : 5

            // Find the best contiguous block of expectedCount that fits the range
            var bestMatch: [DetectedNumber]? = nil
            var bestMatchInvalidCount = Int.max

            for i in 0...(colNumbers.count - expectedCount) {
                let sub = Array(colNumbers[i..<(i + expectedCount)])
                let invalidCount = sub.filter { !columnRanges[c].contains($0.value) }.count
                if bestMatch == nil || invalidCount < bestMatchInvalidCount {
                    bestMatch = sub
                    bestMatchInvalidCount = invalidCount
                }
            }

            guard let match = bestMatch else { return .incomplete }

            var targetIdx = 0
            for r in 0..<5 {
                if r == 2 && c == 2 {
                    finalGrid[r][c] = "" // FREE
                } else {
                    finalGrid[r][c] = String(match[targetIdx].value)
                    targetIdx += 1
                }
            }
        }

        // 4. Final validation of ranges
        for c in 0..<5 {
            for r in 0..<5 {
                if r == 2 && c == 2 { continue }
                guard let num = Int(finalGrid[r][c]) else { return .incomplete }
                if !columnRanges[c].contains(num) {
                    return .invalidRange(number: num, column: columnLabels[c])
                }
            }
        }

        return .success(grid: finalGrid)
    }

    // MARK: - Private

    private func extractNumbers(from observations: [VNRecognizedTextObservation], imageSize: CGSize) -> [DetectedNumber] {
        var candidates: [DetectedNumber] = []

        for observation in observations {
            // Get the top candidate text
            guard let topCandidate = observation.topCandidates(1).first else { continue }

            let text = topCandidate.string.trimmingCharacters(in: .whitespacesAndNewlines)
            let digitsOnly = text.filter { $0.isNumber }

            guard let number = Int(digitsOnly), number >= 1, number <= 75 else { continue }

            // Vision boundingBox is normalized [0,1] with bottom-left origin.
            // Convert to absolute coordinates with top-left origin for our spatial algorithm.
            let bbox = observation.boundingBox
            let absX = bbox.midX * imageSize.width
            let absY = (1.0 - bbox.midY) * imageSize.height // invert Y for top-left origin
            let absHeight = bbox.height * imageSize.height

            candidates.append(DetectedNumber(value: number, x: absX, y: absY, height: absHeight))
        }

        return candidates
    }
}

// MARK: - Array Extension

private extension Array where Element == CGFloat {
    func median() -> CGFloat {
        guard !isEmpty else { return 0 }
        let sorted = self.sorted()
        let mid = sorted.count / 2
        if sorted.count % 2 == 0 {
            return (sorted[mid - 1] + sorted[mid]) / 2.0
        } else {
            return sorted[mid]
        }
    }
}
